installers/msi-language/StateDeploy/CustomAction.cs

```diff
@@ -122,21 +123,62 @@ private static ActionResult _installStateTool(Session session, out string stateT
             string zipURL = stateURL + info.version + "/" + paths.ZipFile;
             session.Log(string.Format("Downloading zip file from URL: {0}", zipURL));
             Status.ProgressBar.StatusMessage(session, "Downloading State Tool...");
-            try
+
+            var tokenSource = new CancellationTokenSource();
+            var token = tokenSource.Token;
+
+            Task incrementTask = Task.Run(() =>
             {
-                RetryHelper.RetryOnException(session, 3, TimeSpan.FromSeconds(2), () =>
+                if (token.IsCancellationRequested)
                 {
-                    var client = new WebClient();
-                    client.DownloadFile(zipURL, zipPath);
-                });
+                    token.ThrowIfCancellationRequested();
+                }
+                for (int i = 0; i <= 50; i++)
+                {
+                    if (token.IsCancellationRequested)
+                    {
+                        token.ThrowIfCancellationRequested();
+                    }
+                    Status.ProgressBar.Increment(session, 1);
+                    Thread.Sleep(150);
+                }
+            });
+
+            Task<ActionResult> downloadTask = Task.Run(() =>
+            {
+                try
```

Comment:
Shouldn't this task request the cancellation on the token somewhere?

I assume that otherwise we always need to wait for the progressbar to finish, even if the download has finished before, no?

Reply:
A different way to implement this could also be [Task.WhenAny](
https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenany?view=netframework-4.5). When that returns, you can just always cancel the progress bar task, and wait for the download task.

Reply:
From how I understand it, inspecting the `task.Result` is equivalent to a wait. From [this link](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task-1.result?view=netcore-3.1)
> Accessing the property's get accessor blocks the calling thread until the asynchronous operation is complete; it is equivalent to calling the Wait method.

So we wait for the download task to finish by inspecting the results and then cancel the progress bar. Though I could be misunderstanding how this works as this is the first time I've used threading in C#. Let me know if that makes sense or if there is something I've overlooked.

Reply:
That is interesting.  I did not know that, or expected that to be the case... 

But what I still don't understand is how the cancellation would work: From what I understand the token(source) is never cancelled.  So, I suppose we will always block in line 170 until the progressbar has been updated 50 times... Or did I miss something?

Reply:
Okay I now see what you mean. I forgot to add the `token.Cance()` call. Further to that it would still have blocked on the wait call. I've updated now and tested to ensure that the progress bar incrementing is cancelled when the download has finished.